src/data_info/services/path_lists.clj

```diff
@@ -0,0 +1,124 @@
+    (concat [data-item] sub-listing)))
+
+(defn- keep-top-level-file?
+  "A filter predicate to keep only files with an info-type that matches one in the given `info-types` list,
+   or to keep all files if `info-types` is empty."
+  [info-types]
+  (if (empty? info-types)
+    (constantly true)
+    (comp (partial some (set info-types)) list :infoType)))
+
+(defn- label-matches?
+  "Returns true if the given `name-pattern` matches the given label, or if the given `name-pattern` is blank."
+  [name-pattern label]
+  (or (string/blank? name-pattern)
+      (re-find (re-pattern name-pattern) label)))
+
+(defn- keep-data-item?
+  "Returns true if the given `data-item` has permissions set,
+   its label passes the `label-matches?` check,
+   and if it's a file or folder depending on the given `folders-only?` flag."
+  [name-pattern folders-only? {:keys [label permission] :as data-item}]
+  (and permission
+       (label-matches? name-pattern label)
+       (if (stat-is-dir? data-item)
+         folders-only?
+         (not folders-only?))))
+
+(defn- paths->ht-path-list
+  "Filters the given paths and returns a string of these paths appended to an HT Path List header.
+   Throws an error if the filtering params result in no matching paths."
+  [cm user name-pattern info-types folders-only? recursive? paths]
+  (let [{folder-paths true file-paths false} (group-by (partial item-info/is-dir? cm) paths)
+        files          (->> file-paths
+                            (map (partial stat/path-stat cm user))
+                            (filter (keep-top-level-file? info-types)))
+        folders        (->> folder-paths
+                            (mapcat (partial folder-listing user info-types folders-only? recursive?))
+                            (mapcat (partial list-item-with-subitems user info-types folders-only? recursive?)))
+        filtered-paths (->> (concat files folders)
+                            (filter (partial keep-data-item? name-pattern folders-only?))
+                            (map :path))]
+
+    (when (empty? filtered-paths)
+      (throw+ {:error_code ERR_NOT_FOUND
+               :reason "No paths matched the request."}))
+
+    (string/join "\n" (concat [(cfg/path-list-file-identifier)] filtered-paths))))
+
+(defn create-path-list
+  "Creates an HT Path List file from the given set of paths and filtering params.
+   The resulting HT Path List file will contain only file or only folder paths (depending on the `folders-only` param),
+   but no paths of folders in the initial given list will be included.
+   If `recursive` is true, then all subfolders (plus all their files and subfolders)
+   of any given folder paths are parsed and filtered as well.
+   Throws an error if the filtering params result in no matching paths."
+  [{:keys [user dest name-pattern info-type folders-only recursive]} {:keys [paths]}]
+  (irods/with-jargon-exceptions :client-user user [cm]
+    (validators/user-exists cm user)
```

Comment:
You shouldn't need this, since you're using a client user. If the client user doesn't exist, `with-jargon` will already have failed out with the same error.